### PR TITLE
feat(linear_solve): matrix inversion linear solver with neumann series approximation

### PR DESCRIPTION
**Pull request by Benjamin-eecs**
## Description

matrix inversion linear solver with neumann series approximation

## Types of changes

- [x] New feature (non-breaking change which adds core functionality)
- [x] Documentation (update in the documentation)
- [x] Example (update in the folder of example)

## Checklist

- [x] I have read the [CONTRIBUTION](https://torchopt.readthedocs.io/en/latest/developer/contributing.html) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the code using `make lint` (**required**)
- [x] I have ensured `make test` pass. (**required**)

### CODE REVIEW

File: torchopt/linear_solve/inv.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        The solution with the same shape as ``b``.
+    """
+    dtype = None
+    leaves = pytree.tree_leaves(b)
+    if len(leaves) > 0:
+        dtype = leaves[0].dtype
+    if ridge is not None:
+        matvec = make_ridge_matvec(matvec, ridge=ridge)
+
+    if len(cat_shapes(b)) == 0:  # pylint: disable=no-else-return
```
Comment (XuehaiPan) [inline torchopt/linear_solve/inv.py:87]: No else return.

File: torchopt/linear_solve/inv.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        return b / materialize_array(matvec, cat_shapes(b), dtype=dtype)
+    if len(cat_shapes(b)) == 1:
+        if ns:
+            A = materialize_array(matvec, cat_shapes(b), dtype=dtype)
+            return linalg.ns(A, b, **kwargs)
+        A = materialize_array(matvec, cat_shapes(b), dtype=dtype)
+        return pytree.tree_matmul(torch.linalg.inv(A), b)  # type: ignore
```
Comment (XuehaiPan) [inline torchopt/linear_solve/inv.py:94]: Use specific type ignore.

File: torchopt/linalg/ns.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        if alpha is not None:
+            b = pytree.tree_map(lambda lhs, rhs: lhs - alpha * rhs, b, pytree.tree_matmul(A, b))
+        else:
+            b = pytree.tree_sub(b, pytree.tree_matmul(A, b))
+        inv_A_hat_b = pytree.tree_add(inv_A_hat_b, b)
+    return inv_A_hat_b
+
+
+def ns_inv(
+    A: TensorTree,
```
Comment (XuehaiPan) [inline torchopt/linalg/ns.py:73]: Support callable is better.

**Review by XuehaiPan — CHANGES_REQUESTED** [review]
CHANGELOG

**Comment by Benjamin-eecs** [comment]
> CHANGELOG

added

File: torchopt/pytree.py [hunk 4]
```diff
@@ -48,10 +64,80 @@ def tree_flatten_as_tuple(
...
+
+
+def acc_matmul(*args: T) -> T:
+    """Accumulate matrix multiplication."""
+    return functools.reduce(operator.matmul, args)
+
+
+def tree_add_scalar_mul(tree_x: TensorTree, tree_y: TensorTree, alpha: float = 1.0) -> TensorTree:
+    """Computes tree_x + alpha * tree_y."""
+    return tree_map(lambda x, y: x.add(y, alpha=alpha), tree_x, tree_y)
```
Comment (XuehaiPan) [inline torchopt/pytree.py:84]:
```suggestion
def tree_add_scalar_mul(tree_x: TensorTree, tree_y: TensorTree, alpha: Optional[float] = None) -> TensorTree:
    """Computes tree_x + alpha * tree_y."""
    if alpha is None:
        return tree_map(lambda x, y: x.add(y), tree_x, tree_y)
    return tree_map(lambda x, y: x.add(y, alpha=alpha), tree_x, tree_y)
```

And reorder functions to match `__all__`.

File: torchopt/linear_solve/inv.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    if len(leaves) > 0:
+        dtype = leaves[0].dtype
+    if ridge is not None:
+        matvec = make_ridge_matvec(matvec, ridge=ridge)
+
+    if len(cat_shapes(b)) == 0:
+        return b / materialize_array(matvec, cat_shapes(b), dtype=dtype)
+    if len(cat_shapes(b)) == 1:
+        if ns:
+            # A = materialize_array(matvec, cat_shapes(b), dtype=dtype)
```
Comment (XuehaiPan) [inline torchopt/linear_solve/inv.py:80]: ?

File: torchopt/linear_solve/inv.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+
+__all__ = ['solve_inv']
+
+
+def _solve_inv(
+    matvec: Callable[[TensorTree], TensorTree],  # (x) -> A @ x
+    b: TensorTree,
+    ridge: Optional[float] = None,
+    ns: bool = False,
```
Comment (XuehaiPan) [inline torchopt/linear_solve/inv.py:54]: docs?

File: tests/test_implicit.py [hunk 7]
```diff
@@ -426,7 +427,117 @@ def matvec(u):
...
+        updates, optim_state = optim.update(grads, optim_state)
+        l2reg_torch = torchopt.apply_updates(l2reg_torch, updates)
+
+        xs = jnp.array(xs.numpy(), dtype=np_dtype)
+        ys = jnp.array(ys.numpy(), dtype=np_dtype)
+        xq = jnp.array(xq.numpy(), dtype=np_dtype)
+        yq = jnp.array(yq.numpy(), dtype=np_dtype)
+
+        def outer_level(params_jax, l2reg_jax, xs, ys, xq, yq):
+            w_fit = ridge_solver_jax_cg(params_jax, l2reg_jax, xs, ys)
```
Comment (XuehaiPan) [inline tests/test_implicit.py:540]: cg?

File: tests/test_implicit.py [hunk 8]
```diff
@@ -426,7 +427,117 @@ def matvec(u):
...
+    ns_flag=[True, False],
+)
+def test_rr_solve_inv(
+    dtype: torch.dtype,
+    lr: float,
+    ns_flag: bool,
+) -> None:
+    if dtype == torch.float64:
+        if ns_flag == True:
+            pytest.skip('Neumann Series test skips torch.float64.')
```
Comment (XuehaiPan) [inline tests/test_implicit.py:455]: why?

Comment (Benjamin-eecs) [reply tests/test_implicit.py:455]: cant pass, neumann series approximation has assumption with repect to problems

**Review by XuehaiPan — APPROVED** [review]

**Review by XuehaiPan — APPROVED** [review]

**Review by XuehaiPan — APPROVED** [review]

**Pull request merged**